pkg/kubelet/volumemanager/volume_manager.go

```diff
@@ -389,11 +389,17 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
 
 // getUnattachedVolumes returns a list of the volumes that are expected to be attached but
 // are not currently attached to the node
-func (vm *volumeManager) getUnattachedVolumes(expectedVolumes []string) []string {
+func (vm *volumeManager) getUnattachedVolumes(podName types.UniquePodName, expectedVolumes []string) []string {
 	unattachedVolumes := []string{}
+	desiredVolumes := vm.desiredStateOfWorld.GetVolumesToMount()
```

Comment:
I am also concerned about the performance.

Reply:
I came up with another way to improve the performance.
We should call `vm.actualStateOfWorld.VolumeExistsWithSpecNameSet`  instead of `vm.desiredStateOfWorld.GetVolumesToMount`
create a new function
`func (asw *actualStateOfWorld) VolumeExistsWithSpecNameSet(podName volumetypes.UniquePodName, volumeSpecName sets.String)sets.String`

Reply:
Thanks for suggestions. I wasn't quite sure about the performance hit too. I'll examine the @xiaoanyunfei's suggestion and see if there is some better way to fix this.

Reply:
I did take a look and added a new method `VolumeExistsWithiOuterVolumeSpecName(podName volumetypes.UniquePodName, outerVolumeSpecName string) bool` to ASW. It's inspired by `VolumeExistsWithSpecName` and does basically the same: retrurns `true` when the volume with the "outer" name is attached to the node. And I can use this method to filter the unattached volumes from the volume manager. This way we should not iterate over so many volumes. I'm yet to write some tests for this and will push a new code then.